lib/Transform/IR/PointerScalarizer.cpp

```diff
@@ -0,0 +1,455 @@
+  auto &LI = getAnalysis<LoopInfoWrapperPass>().getLoopInfo();
+  auto &LoopAttr = getAnalysis<LoopAttributesDeductionPass>();
+  auto &AT = getAnalysis<EstimateMemoryPass>().getAliasTree();
+  auto &TLI = getAnalysis<TargetLibraryInfoWrapperPass>().getTLI(F);
+
+  auto MDsToAttach = SmallVector<Metadata*, 8>();
+  for_each_loop(LI, [&TraitPool, &LoopAttr, &AT, &TLI, &F, &MDsToAttach](Loop *L) {
+    // 1. Find memory that was marked anti/flow/output;
+    // 2. Check that this memory is accessed inside the loop;
+    // 3. If possible, copy its value at the preheader and store it back after the exit block(s);
+    // 4. Replace all load/store instructions and their users in the loop's body
+    //    with corresponding operations with the copied memory;
+    // 5. Map inserted instructions with the original value using DI nodes.
+    if (!L->getLoopID() ||
+        !L->getLoopPreheader() ||
+        !LoopAttr.hasAttr(*L, Attribute::NoUnwind) ||
+        LoopAttr.hasAttr(*L, Attribute::ReturnsTwice) ||
+        LoopAttr.hasAttr(*L, AttrKind::AlwaysReturn))
+      return;
+    auto &Pool = TraitPool[L->getLoopID()];
+    if (!Pool)
+      return;
+    SmallDenseSet<Value *> Values;
+    SmallDenseMap<Value *, DIMemoryTrait *> ValueTraitMapping;
+    for (auto &T : *Pool) {
+      if (!T.is_any<trait::Anti, trait::Flow, trait::Output>())
+        continue;
+      for (auto &V : *T.getMemory()) {
+        if (!V.pointsToAliveValue() ||
+            isa<UndefValue>(V) ||
+            isa<GetElementPtrInst>(V) ||
+            isa<LoadInst>(V))
+          continue;
+        if (auto *I = dyn_cast<Instruction>(V))
+          if (L->contains(I))
+            continue;
+        for (auto *User : V->users()) {
+          if (auto *SI = dyn_cast<StoreInst>(User))
+            if (L->contains(SI) && SI->getPointerOperand() == V) {
+              Values.insert(V);
+              ValueTraitMapping[V] = &T;
+              break;
+            }
+        }
+      }
+    }
+    for (auto *Val : Values) {
+      auto DIB = new DIBuilder(*F.getParent());
+      auto Ctx = ScalarizerContext(Val, F, L, DIB);
+      if (hasVolatileInstInLoop(Ctx) ||
+          !analyzeAliasTree(Val, AT, L, TLI)) {
+        ValueTraitMapping.erase(Val);
+        continue;
+      }
+      SmallVector<DIMemoryLocation, 4> DILocs;
+      Optional<DIMemoryLocation> DIM;
+      if (isa<AllocaInst>(Ctx.V)) {
+        DIM = findMetadata(Ctx.V, DILocs);
+      }else {
+        bool HasDbgInLoop = false;
+        for (auto *BB : L->getBlocks())
+          for (auto &I : BB->getInstList())
+            if (auto *DVI = dyn_cast<DbgValueInst>(&I))
+              if (DVI->getValue() == Ctx.V) {
+                HasDbgInLoop = true;
+                break;
+              }
+        if (!HasDbgInLoop)
+          DIM = findMetadata(Ctx.V, {&L->getHeader()->front()}, AT.getDomTree(), DILocs);
+      }
+      if (!DIM.hasValue() || !DIM->Var || !DIM->Loc)
+        continue;
```

Comment:
- Глобальные переменные нужно обрабатывать также как и `alloca`.
- `Loc` может и не быть, например, для глобальных переменных ее нет.

Соответствующим образом надо поправить `createDbgInfo`, иначе сейчас глобальные переменные не преобразуются.

```suggestion
      if (isa<AllocaInst>(Ctx.V) || isa<GlobalVariable>(Ctx.V)) {
        DIM = findMetadata(Ctx.V, DILocs);
      }else {
        bool HasDbgInLoop = false;
        for (auto *BB : L->getBlocks())
          for (auto &I : BB->getInstList())
            if (auto *DVI = dyn_cast<DbgValueInst>(&I))
              if (DVI->getValue() == Ctx.V) {
                HasDbgInLoop = true;
                break;
              }
        if (!HasDbgInLoop)
          DIM = findMetadata(Ctx.V, {&L->getHeader()->front()}, AT.getDomTree(), DILocs);
      }
      if (!DIM || !DIM->isValid())
        continue;
```

Reply:
Не проходит `8` тестов из `tests/analysis/da_di`.